test/end2end_test.go

```diff
@@ -5284,6 +5284,53 @@ func (s) TestGRPCMethod(t *testing.T) {
 	}
 }
 
+// renameProtoCodec is an encoding.Codec wrapper that allows customizing the
+// Name() of another codec.
+type renameProtoCodec struct {
+	encoding.Codec
+	name string
+}
+
+func (r *renameProtoCodec) Name() string { return r.name }
+
+func (s) TestForceCodecName(t *testing.T) {
+	wantContentTypeCh := make(chan string, 1)
+	defer close(wantContentTypeCh)
+
+	ss := &stubserver.StubServer{
+		EmptyCallF: func(ctx context.Context, in *testpb.Empty) (*testpb.Empty, error) {
+			md, ok := metadata.FromIncomingContext(ctx)
+			if !ok {
+				return nil, status.Errorf(codes.Internal, "no metadata in context")
+			}
+			if got, want := md["content-type"], <-wantContentTypeCh; len(got) != 1 || got[0] != want {
+				return nil, status.Errorf(codes.Internal, "got content-type=%q; want [%q]", got, want)
+			}
+			return &testpb.Empty{}, nil
+		},
+	}
+	protoCodec := encoding.GetCodec("proto")
+	if err := ss.Start([]grpc.ServerOption{grpc.ForceServerCodec(encoding.GetCodec("proto"))}); err != nil {
+		t.Fatalf("Error starting endpoint server: %v", err)
+	}
+	defer ss.Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	codec := &renameProtoCodec{Codec: protoCodec, name: "some-test-name"}
+	wantContentTypeCh <- "application/grpc+some-test-name"
```

Comment:
Nit: on potential blocking sends wait on a timeout context (or a timer) with defaultTestShortTimeout too so tests don't potentially block for 10 minutes if behavior change within codebase/test is flaky.

Reply:
I know probably this will probably never timeout on this send, but I like always accounting for the possibility.